prog/dftb+/lib_math/eigensolver.F90

```diff
@@ -2042,4 +2054,368 @@ subroutine dblecmplx_zhbgv(ab, bb, w, uplo, z)
 
   end subroutine dblecmplx_zhbgv
 
+#:if WITH_GPU
+
+  subroutine real_magma_ssygvd(ngpus, a, b, w, uplo, jobz, itype)
+
+    !> contains the matrix for the solver, returns eigenvectors if requested (matrix always
+    !> overwritten on return anyway)
+    real(rsp), intent(inout) :: a(:,:)
+    real(rsp), allocatable :: H(:)
+
+    !> contains the second matrix for the solver (overwritten by Cholesky factorization)
+    real(rsp), intent(inout) :: b(:,:)
+    real(rsp), allocatable :: S(:)
+
+    !> eigenvalues
+    real(rsp), intent(out) :: w(:)
+
+    !> upper or lower triangle of the matrix
+    character, intent(in) :: uplo
+
+    !> compute eigenvalues 'N' or eigenvalues and eigenvectors 'V'
+    character, intent(in) :: jobz
+
+    !> optional specifies the problem type to be solved 1:A*x=(lambda)*B*x, 2:A*B*x=(lambda)*x,
+    !> 3:B*A*x=(lambda)*x default is 1
+    integer, optional, intent(in) :: itype
+
+    real(rsp), allocatable :: work(:)
+    integer :: lwork
+    integer, allocatable :: iwork(:)
+    integer :: liwork
+
+    integer n, info, iitype
+   
+    integer :: ngpus
+    integer :: ldm, nwarp, i, i0    
+
+    @:ASSERT(uplo == 'u' .or. uplo == 'U' .or. uplo == 'l' .or. uplo == 'L')
+    @:ASSERT(jobz == 'n' .or. jobz == 'N' .or. jobz == 'v' .or. jobz == 'V')
+    @:ASSERT(all(shape(a)==shape(b)))
+    @:ASSERT(all(shape(a)==size(w,dim=1)))
+    n=size(a,dim=1)
+    @:ASSERT(n>0)
+    if (present(itype)) then
+      iitype = itype
+    else
+      iitype = 1
+    end if
+    @:ASSERT(iitype >= 1 .and. iitype <= 3 )
+
+    !> Padding for shared memory to avoid bank conflicts, 32-bit mode
+    nwarp = 32      
+    ldm   = nwarp* ((n+nwarp-1)/nwarp)
+    allocate(H(ldm*ldm), S(ldm*ldm),stat=info) 
+    if (info.ne.0) call  error('CPU alloc error in real_magma_ssygvd. Exiting')
+    
+    !> Convert nxn matrices to 1D arrays
+    i0=1
+    do i=1,n 
```

Comment:
Fortran intrinsic reshape probably faster, might be simpler code too.

Reply:
I didn't change this because although is a nxn matrix conversion to a 1-dimensional one, we usually pad the number of rows to a multiple of 32 on the GPU.